controllers/mesh.py

```diff
@@ -1,11 +1,132 @@
-from flask import Blueprint, jsonify, request
+import os
+import time
+from flask import Blueprint, jsonify, request, url_for, flash
 from flask_security import login_required, current_user
 from models.mesh import Mesh
+from models import db
 from odk2odm import odk_requests
 from odk2odm import odm_requests
+from celery import states
+from celery.exceptions import Ignore
 # API components that retrieve or download data from database for use on front end
+from cryptography.fernet import Fernet
+
 mesh_api = Blueprint("mesh_api", __name__)
 
+from app import celery
+
+@celery.task(bind=True)
+def test_job(self):
+    print("Running job as a Thread!")
+    for i in range(10):
+        print(f"Processing {i}/10")
+        self.update_state(
+            state='PROGRESS',
+            meta={
+                'current': i,
+                'total': 10,
+                'status': "Working..."
+            }
+        )
+        time.sleep(1)
+    return {
+        'current': 100,
+        'total': 100,
+        'status': 'Job completed!',
+        'result': 42
+    }
+
+
+@celery.task(bind=True)
+def odk2odm(self, submissions=[], odk={}, odm={}):
+    def transfer(att, state="PROGRESS"):
+        """
+        subroutine to transfer one file
+        :param att: attachment details from odk server
+        :return:
+        """
+        print(odm)
+        print(f"Transfer of {att}")
+        res = odm_requests.get_thumbnail(filename=att["name"], **odm)
+        if res.status_code == 200:
+            msg = f"{att['name']} already uploaded, so skipping..."
+            print(msg)
+            self.update_state(
+                state='PROGRESS',
+                meta={
+                    'current': n,
+                    'total': len(submissions),
+                    'status': msg
+                }
+            )
+            return
+        # retrieve photo
+        res = odk_requests.attachment(instanceId=sub["instanceId"], filename=att["name"], **odk)
+        try:
+            if res.status_code != 200:
+                # something went wrong, so report that
+                raise IOError(res.json())
+        except:
+            raise IOError(f"Cannot reach ODK server at {odk['url']}")
+        # post it on the task
+        # retrieve file contents (should only contain "images" but checking for all keys to make sure)
+        fields = {"images": (att["name"], res.content, "images/jpg")}
+        try:
+            res = odm_requests.post_upload(fields=fields, **odm)
+            if res.status_code == 200:
+                status = f"{att['name']} uploaded..."
+            else:
+                raise IOError(res.json())
+        except:
+            raise IOError(f"Cannot reach ODM server at {odm['url']}")
```

Comment:
Nice error handling. Covering lots of different cases and informative failure messages.